mgmt-vrf-design-doc.md

```diff
@@ -0,0 +1,313 @@
+                                          Dest IP = Front end port IP                      Dest IP ≠ Front end port IP
+
+     Packet coming via mgmt. port         Not applicable                                   Not applicable
+
+     Packet coming via Front end port	  All IP services listed in Table 1 should    	   Treat the packet like any other data traffic.
+                                          be supported                                     To be sent out as per the routing table lookup
+                                                                                           in management VRF
+     3.Switch originated traffic
+      -------------------------
+
+      Unless explicitly specified to use mamangment VRF, all the services will use the default routing table to egress out of switch.
+
+Design Approach
+---------------
+
+    | Features         | Approach  Name space            | Approach  L3MDev                             |
+    |------------------|----------------------------------|-----------------------------------------------|
+    | Kernel Support   | Yes                              | Yes                                           |
+    | Scalability      | Less scalable                    | Better Scalable                               |
+    | Protocol support | IP services should be replicated | IP services can be enslaved to a L3 interface |
+    | Performance      | Overhead  of replicated services | Shared services                               |
+
+Network namespaces were designed to provide a complete isolation of the entire network stack —devices, network addresses, neighbor and routing tables, protocol ports and sockets. 
+VRF On the other hand is a network layer 3 feature and as such should really only impact FIB lookups. While the isolation provided by a namespace includes the route tables, a network namespace will have  a significant overhead as we will end up having two network stacks.
+
+Using network name space as VRF’s will involve having services replicated across all VRF’s – for example we will have run two instances of lldp/ssh for Management and Default VRF’s respectively. 
+Also in terms of extensibility, using namespaces for Multi VRF means N-VRF’s which will have N-application instances and also the additional complexity of aggregating all the data across these instances.
+ 
+With L3-Mdev on 4.9 , VRF’s are created and enslaved to an L3 Interface and services like ssh can be shared across VRF’s thereby avoiding overhead of running multiple instances of IP services.
+
+
+Phase-1
+-------
+
+In this Phase mgmt-vrf feature will be enabled by default when SONiC boots with Debian 4.9 kernel, configuration of mgmt-vrf will be
+done using linux commands.
+
+
+DB Schema
+---------
+
+The config_db.json schema is updated to have the vrfname in MGMT_INTERFACE. The schema representation will look like below
+
+    "MGMT_INTERFACE": {
+        "eth0|10.11.150.19/24": {

```

Comment:
Please outline in your design document how this will work when eth0 ip address and gateway are obtained via dhcp and not statically configured.

Reply:
will update